### PR TITLE
Walkthrough 3 - low code development

### PR DESCRIPTION
**Pull request by pwright**

### CODE REVIEW

**Comment by damienomurchu** [comment]
@fin will review this when I've a cluster up with the latest versions of the integration products

**Comment by pwright** [comment]
@aidenkeating  do we need to update the walkthrough.json in this PR?

**Comment by aidenkeating** [comment]
@finp No we shouldn't need to, the `.templates` config is optional. Happy to test this out on OpenShift 4 if needed though?

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+
+. Click *Validate* in Fuse Online and verify that a *Slack has been successfully validated.* message appears. Click *Next*.
+
+. Enter the following in the *Connection Name* field on the *Add Connection Details* screen:
++
+----
+Greeting Slack Target
+----
+
+. Click *Create*.
```
Comment (damienomurchu) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
. Click *Save*.
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+=== Creating an Integration with API details
+
+In this procedure, you create an integration with an API and include a data type.
+A *Data Type* represents a data structure that can be passed to your API and can then be used as parameters in your API definition and within the Fuse Online *Integration Flow Editor* to transform and map data as part of an *Integration*.
+
+. Click the *Create Integration* button to start the *New Integration* wizard.
+
+. Choose *API Provider* on the subsequent *Choose a Start Connection* screen.
+
+. When prompted choose *Create from scratch* and click *Next* to navigate to the *API Designer*.
```
Comment (damienomurchu) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
. When prompted choose *Create* and click *Next* to navigate to the *API Designer*.
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+.. Click *No description* beside the *200 OK* response and enter the following in the *Description* field:
++
+----
+Greeting response
+----
+
+. Click *Save* in the top right corner of the UI to save your work and be directed back to the *New Integration* wizard.
+
+. Click *Next* on the *New Integration* wizard screen.
+
```
Comment (damienomurchu) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
. Click *Create flow* for the /greeting POST operation you just created
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+.. Click *Add*.
+.. Click *No description* beside the *200 OK* response and enter the following in the *Description* field:
++
+----
+Greeting response
+----
+
+. Click *Save* in the top right corner of the UI to save your work and be directed back to the *New Integration* wizard.
+
+. Click *Next* on the *New Integration* wizard screen.
```
Comment (damienomurchu) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
. Click *Publish* on the *New Integration* wizard screen.
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+----
+Greeting Integration
+----
+.. Enter the following in the *Description* field:
++
+----
+my greeting api
+----
+
+. Click *Save and Continue* to save your API design.
```
Comment (damienomurchu) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
. Click *Save* to save your API design.
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+[time=10]
+== Implementing and publishing the API
+
+. Log in to the link:{fuse-url}[Red Hat Fuse Online, window="_blank"] console.
+
+. Select *Integrations* from the left hand menu. 
+
+. Choose *Edit* from the menu for the *Greeting Integration* entry. The *Operations* page appears.
+
+. Click the *Edit flow* button. The *Add to Integration* page appears displaying the *Provided API* and the *Provided API Return Path*.
```
Comment (damienomurchu) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
. Click the *Create flow* button. The *Add to Integration* page appears displaying the *Provided API* and the *Provided API Return Path*.
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+
+. Click the blue plus icon to add a step.
+The left hand side of the *Choose a connection* page lists the steps in your *Integration*.
+
+. Select your *Greeting Slack Target*.
+
+. When prompted to *Choose an action* select *Channel*. You can use this to send a message to a specific channel in your Slack workspace.
+
+. Use the *Channel* dropdown to select the *low-code-demo* channel and click *Done*.
```
Comment (damienomurchu) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
. Use the *Channel* dropdown to select the *low-code-demo* channel and click *Next*.
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+. Choose 'i-greeting-api' from the *Name* list.
+. Click *Create Service*.
++
+This process can take a few minutes.
+
+// The 3scale dashboard is displayed, with notification: `The service will be imported shortly. You will receive a notification when it is done.`
+
+. Edit the API:
+
+.. From the *Overview* screen, select the *Configure APIcast* button.
```
Comment (damienomurchu) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
.. From the *Overview* screen, select the *edit APIcast configuration* link.
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+
+----
+
+Did the message "Hello from John" appear in your Slack channel?
+****
+
+[type=verificationFail]
+Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.
+
+=== Invoking the API
```
Comment (damienomurchu) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc:462]: This section seems to duplicate the previous one:

Screenshot from 2019-08-22 15-22-57

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 10]
```diff
@@ -0,0 +1,465 @@
...
+----
+.. Select *Set Custom Key*.
+
+
+[type=verification]
+****
+Use an HTTP client to invoke the route, for example:
+
+----
+curl -d '{"name":"John"}' -H "Content-Type: application/json" -X POST "{route}/greeting?user_key=test"
```
Comment (damienomurchu) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]: `{route}` renders as {route} in walkthrough, and isn't replaced by actual route

**Comment by damienomurchu** [comment]
@finp Finished verifying walkthrough in a cluster provisioned with 1.5.0-RC2. Was able to complete walthrough, but there have been some small changes in the UI's of Fuse & 3scale and have left some suggested corrections. Also some slight duplication in last step which I flagged in a comment too.

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+:3scale-version: 7.2
+:fuse-url: https://eval.apps.city.openshiftworkshop.com/
+:3scale-url: https://eval.apps.city.openshiftworkshop.com/
+:route: https://wt2-{user-sanitized-username}-3scale.{openshift-app-host}
+
+[id='no-code-container-native-api-development']
+
+= {walkthrough}
+
+This walkthrough demonstrates how to implement a container-native OpenAPI RESTful API without coding.
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc:15]: Formatting looks a bit iffy.  See https://www.dropbox.com/s/s6jfzqduojpk1mq/Screenshot%202019-08-23%20at%2011.46.27.png?dl=0

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+
+Container-native applications::
+Applications that are designed to run in a containerised environment.
+
+// This is taken right from https://github.com/OAI/OpenAPI-Specification
+OpenAPI Specification::
+A standard, programming language-agnostic interface description for REST APIs, which allows both humans and computers to discover and understand the capabilities of a service without requiring access to source code, additional documentation, or inspection of network traffic. When properly defined via OpenAPI, a consumer can understand and interact with the remote service with a minimal amount of implementation logic.
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]: Is this sentence too long?

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+=== Creating a Slack Workspace
+
+.Prerequisite
+You already have an account with link:https://slack.com/[Slack].
+
+.Procedure
+. Navigate to the link:https://slack.com/create[Create a new workspace, window="_blank"] page.
+
+. Enter your email address and click *Next*.
+
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc:68]: If you have 2FA enabled on your slack account, you will have to enter the 2FA code after clicking *Next*.  It won't go straight to "What's the name of your company or team?"

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+. When prompted *What's the name of your company or team'* enter::
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
. When prompted *What's the name of your company or team?* enter::
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+----
+
+. When prompted *What’s a project your team is working on?* enter:
++
+----
+low-code-demo
+----
+
+. Click *Skip for now* when prompted to add team members.
+
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc:82]:
```suggestion
. Click *See Your Channel in Slack*
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 16]
```diff
@@ -0,0 +1,458 @@
...
+Can you access your workspace, and navigate to the *low-code-demo* channel listed on the left of the workspace?
+
+
+[type=verificationFail]
+Verify that you followed each step in the procedure above.  If you are still having issues, check your email, there may be information sent to you about accessing your workspace.
+
+
+=== Creating the Slack Connection in Fuse Online
+
+. Log in to the link:{fuse-url}[Red Hat Fuse Online, window="_blank"] console.
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc:95]: {fuse-url} points to https://eval.apps.city.openshiftworkshop.com/ , which is a dead link.  Unable to proceed further.

**Review by briandooley — CHANGES_REQUESTED** [review]
Unable to get past fuse app section.

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 17]
```diff
@@ -0,0 +1,458 @@
...
+. Choose *Connections* from the left navigation menu.
+
+. Click the *Create Connection* button in the top right corner.
+
+. Choose *Slack* from the list of displayed *Connectors*. 
+You are prompted for configuration details.
+
+. Obtain a URL for the *Slack Webhook URL* field:
+.. Navigate to the Slack Apps portal at link:https://api.slack.com/apps[api.slack.com/apps].
+.. Login using the email address associated with your workspace.
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]: You should be already logged into slack at this stage.

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 18]
```diff
@@ -0,0 +1,458 @@
...
+.. In the *Post to* field select *low-code-demo* and click *Allow*.
++
+You are redirected back to the *Incoming Webhooks* page. 
+.. Copy the *Webhook URL* displayed at the bottom of the page using the *Copy* button. 
+
+. Paste the copied *Webhook URL* value into the *Slack Webhook URL* field in Fuse Online.
+
+. Obtain a value for the *Token for Accessing Slack API* field:
+.. Navigate to the Slack API Token portal at link:https://api.slack.com/custom-integrations/legacy-tokens[api.slack.com/custom-integrations/legacy-tokens, window="_blank"].
+.. Login using the email address associated with your workspace.
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]: Logged in already.

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 19]
```diff
@@ -0,0 +1,458 @@
...
+.. Scroll down to the *Legacy token generator* section.
+.. Click *Create token* button next to a workspace you own, or have admin rights for.
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
.. Click *Create token* button next to *low-code-workspace*.
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 20]
```diff
@@ -0,0 +1,458 @@
...
+Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.
+
+[time=10]
+== Creating an API Integration in Fuse Online
+
+=== Creating an Integration with API details
+
+In this procedure, you create an integration with an API and include a data type.
+A *Data Type* represents a data structure that can be passed to your API and can then be used as parameters in your API definition and within the Fuse Online *Integration Flow Editor* to transform and map data as part of an *Integration*.
+
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc:170]:
```suggestion

. Choose *Integrations* from the left navigation menu.

```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 21]
```diff
@@ -0,0 +1,458 @@
...
+. In the window that appears, enter the following in the *Path* field:
++
+----
+/greeting
+----
+
+. Click *Add* to confirm your entry.
+
+. Click your new */greeting* endpoint under the *Paths* heading.
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc:234]: No need for this instruction, already clicked.

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 22]
```diff
@@ -0,0 +1,458 @@
...
+----
+Name
+----
+
+. Enter the following JSON in the *Enter JSON Example* field:
++
+[subs="attributes+"]
+----
+{
+    "name": "shadowman"
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc:206]:
```suggestion
"name": "shadowman" 
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 23]
```diff
@@ -0,0 +1,458 @@
...
+Greet with name
+----
+
+. Enter the following in the *Operation ID* field:
++
+----
+greetname
+----
+
+. In the *Request Body* section, choose the *Name* type that you created earlier.
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]: Needs more details.  It is not possible to choose the *Name* data type as "No request body has been defined.".  Type should be "string as string".

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 24]
```diff
@@ -0,0 +1,458 @@
...
+.. Click *Add*.
+.. Click *No description* beside the *200 OK* response and enter the following in the *Description* field:
++
+----
+Greeting response
+----
+
+. Click *Save* in the top right corner of the UI to save your work and be directed back to the *New Integration* wizard.
+
+. Click *Publish* on the *New Integration* wizard screen.
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc:271]: There is no "publish" button.  Clicking "Next" brings you to a screen with a "Publish" button.

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 25]
```diff
@@ -0,0 +1,458 @@
...
+----
+Greeting Integration
+----
+.. Enter the following in the *Description* field:
++
+----
+my greeting api
+----
+
+. Click *Save* to save your API design.
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
. Click *Save and Publish* to save your API design.
```

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc [hunk 26]
```diff
@@ -0,0 +1,458 @@
...
+
+. Enter the following in the *string* field under the dropdown:
++
+----
+Hello from 
+----
+
+. Click *Done*.
+
+. Click *Publish* to trigger an link:https://docs.openshift.com/container-platform/3.11/creating_images/s2i.html[Source to Image (S2I), window="_blank"] build.
```
Comment (briandooley) [inline walkthroughs/3a-low-code-api-development/walkthrough.adoc]:
```suggestion
. Click *Publish*, and then *Save and Publish", to trigger an link:https://docs.openshift.com/container-platform/3.11/creating_images/s2i.html[Source to Image (S2I), window="_blank"] build.
```

**Review by briandooley — CHANGES_REQUESTED** [review]
Unable to complete walkthrough due to issue detailed in https://issues.jboss.org/browse/INTLY-3021

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc (hunk 22 above)
Comment (damienomurchu) [reply walkthroughs/3a-low-code-api-development/walkthrough.adoc:206]: I would leave this as-is; indentation is expected in the json object.

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc (hunk 11 above)
Comment (pwright) [reply walkthroughs/3a-low-code-api-development/walkthrough.adoc:15]: create a jira

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc (hunk 12 above)
Comment (pwright) [reply walkthroughs/3a-low-code-api-development/walkthrough.adoc]: make suggestion

Comment (briandooley) [reply walkthroughs/3a-low-code-api-development/walkthrough.adoc]: A standard, programming language-agnostic interface description for REST APIs. It allows both humans and computers to discover and understand the capabilities of a service without requiring access to source code, additional documentation, or inspection of network traffic.

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc (hunk 11 above)
Comment (briandooley) [reply walkthroughs/3a-low-code-api-development/walkthrough.adoc:15]: Done.  https://issues.jboss.org/browse/INTLY-3067

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc (hunk 23 above)
Comment (pwright) [reply walkthroughs/3a-low-code-api-development/walkthrough.adoc]: out of date

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc (hunk 25 above)
Comment (pwright) [reply walkthroughs/3a-low-code-api-development/walkthrough.adoc]: done

File: walkthroughs/3a-low-code-api-development/walkthrough.adoc (hunk 26 above)
Comment (pwright) [reply walkthroughs/3a-low-code-api-development/walkthrough.adoc]: done

**Pull request merged**